### PR TITLE
Minimal Varianсe Sampling booster

### PR DESCRIPTION
**Pull request by kruda**
Implementation of minimal variance sampling for stochastic gradient boosting. Contributes to #2644.
**Note for reviers**

I've tested it on examples for binary and multiclass classification from lightgbm repository, tried to train on Higgs dataset using slightly modified config from https://github.com/guolinke/boosting_tree_benchmarks.git. Trained several times on https://www.kaggle.com/c/DontGetKicked .
On tested datasets this MVS implementation shows better score and overfitts, than simple bagging strategy with same sampling rate.
Update:
Mean relative error change table:
|Algorithm\sampling fraction|0.05|0.1625|0.275|0.3875|0.5|0.6125|0.725|0.8375|0.95|
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
|mvs|8.314|3.032|0.775|0.061|-0.391|-0.403|-0.26|-0.185|0.084|
|sgb|13.037|4.516|3.169|2.569|2.105|1.521|1.08|0.343|-0.155|
|goss|17.263|4.672|3.285|2.223|1.196|0.488|0.007|-0.05|-0.079|
|mvs_adaptive|8.314|2.919|0.904|-0.195|-0.451|-0.466|-0.274|-0.079|0.206|

Example on adult dataset:
изображение

Code to reproduce:
https://github.com/kruda/lightgbm_mvs_experiments

### CODE REVIEW

File: R-package/src/Makevars.in
Comment (jameslamb) [inline R-package/src/Makevars.in:29]:
```suggestion
    boosting/mvs.o \
```

Could you please  add a space before the `\` here? Just to make this a bit easier to read.

File: R-package/src/Makevars.win.in
Comment (jameslamb) [inline R-package/src/Makevars.win.in:30]:
```suggestion
    boosting/mvs.o \
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks very much for the contribution! I'll do a more complete review in the next day or two.

For now, I left two very minor suggestions for the R side.

Could you also please add some tests on this feature? As a start, you could try updating the Dask tests to cover this code: https://github.com/microsoft/LightGBM/blob/a421217e4e5f15e8aa5380061ff5d153188cab62/tests/python_package_test/test_dask.py#L44

If you get stuck, please let me know and I'd be happy to help.

**Comment by kruda** [comment]
@jameslamb
I'm sorry for being annoying. Thank you for the tip!

File: include/LightGBM/utils/array_args.h
Comment (StrikerRUS) [inline include/LightGBM/utils/array_args.h:107]: @kruda Could you please extract corresponding to the #4272 part into a separate PR? This will allow to faster merge this fix and not wait for MVS review.
As a bonus, doing so also will remove you from "first contributor" group and help to iterate thought CI cycles faster🙂 .

Comment (kruda) [reply include/LightGBM/utils/array_args.h:107]: Thank you for suggestion! I've created a new pull request for that. https://github.com/microsoft/LightGBM/pull/4280

File: R-package/src/Makevars.win.in
Comment (kruda) [reply R-package/src/Makevars.win.in]: Accepted in 8a01fb8

File: R-package/src/Makevars.in
Comment (kruda) [reply R-package/src/Makevars.in]: Accepted in d499d15

**Comment by kruda** [comment]
@jameslamb do you have any progress with more complete review?

**Review by jameslamb — COMMENTED** [review]
Changes look ok to me! Thanks very much for adding tests and documentation.

I don't know MVS or C++ well enough for my review of this major feature addition to count towards a merge, so just leaving a "comment" review.

File: src/boosting/mvs.cpp
Comment (shiyu1994) [inline src/boosting/mvs.cpp:66]: `need_re_bagging_` should be set to `false` if the `Bagging` is actually carried out, to make it consistent with the behavior in the default random sampling. Please check the code snippet below.
https://github.com/microsoft/LightGBM/blob/08c38efc186dd4a7dcc27574d08207277aa12d54/src/boosting/gbdt.cpp#L228-L233

File: src/boosting/mvs.hpp
Comment (shiyu1994) [inline src/boosting/mvs.hpp:55]: To keep the logic consistent with ordinary `Bagging`, `need_re_bagging` should be set `true` if the configuration for bagging is changed. So we'd better set `need_re_bagging=true` here if either `mvs_lambda_` or `mvs_adaptive_` is changed. (If `mvs_adaptive_` is kept as `true`, but `mvs_lambda_` is changed, then we can ignore this change, though).

File: include/LightGBM/config.h
Comment (shiyu1994) [inline include/LightGBM/config.h:324]:
```suggestion
  // desc = used in MVS boosting. If ``mvs_adaptive == true`` than this value is ignored
```

File: include/LightGBM/config.h
Comment (shiyu1994) [inline include/LightGBM/config.h:335]:
```suggestion
  // desc = used in MVS boosting. If training dataset size is greater than ``mvs_max_sequential_size``, than threshold
```

File: docs/Parameters.rst
Comment (shiyu1994) [inline docs/Parameters.rst:355]:
```suggestion
   -  used in MVS boosting. If training dataset size is greater than ``mvs_max_sequential_size``, than threshold
```

File: docs/Parameters.rst
Comment (shiyu1994) [inline docs/Parameters.rst:343]:
```suggestion
   -  used in MVS boosting. If ``mvs_adaptive == true`` than this value is ignored
```

**Review by shiyu1994 — COMMENTED** [review]
@kruda. Thanks for your great work. I've checked the algorithm. I have no problem about the logic for the `mvs`. Just left a few questions and comments which may require minor modification.

**Comment by StrikerRUS** [comment]
@kruda Thank you very much for writing low-level cpp tests and including `mvs` into parametrizations of Dask tests! This is awesome! But unfortunately, environment coverage of those tests are quite low. Could you please write a standalone basic MVS-specific test as well? For example, please refer to 
https://github.com/microsoft/LightGBM/blob/e35ed5f6e53f8ee81031061a5f8c80c843c09e14/tests/python_package_test/test_engine.py#L2418
https://github.com/microsoft/LightGBM/blob/e35ed5f6e53f8ee81031061a5f8c80c843c09e14/tests/python_package_test/test_engine.py#L2437
https://github.com/microsoft/LightGBM/blob/e35ed5f6e53f8ee81031061a5f8c80c843c09e14/tests/python_package_test/test_engine.py#L1494
or
https://github.com/microsoft/LightGBM/blob/e35ed5f6e53f8ee81031061a5f8c80c843c09e14/tests/python_package_test/test_engine.py#L75

File: docs/Parameters.rst
Comment (shiyu1994) [reply docs/Parameters.rst:343]: BTW, `than` should be `then`, could your please fix this together. Sorry for missing the typo.

**Comment by StrikerRUS** [comment]
Kindly ping @kruda

**Comment by shiyu1994** [comment]
@kruda I can help to manage the details in the comments I left.

**Comment by jameslamb** [comment]
@kruda are you still interested in pursuing this pull request?

**Comment by jameslamb** [comment]
@shiyu1994 what does it mean that this PR is listed in #4677 with your name next to it?

Does it mean you're planning a separate implementation of MVS from your own branch? And if that's it, should this PR be closed? It has been almost 4 months since it received a new commit.

**Comment by shiyu1994** [comment]
Hi @jameslamb, sorry if that confuses you. No. We are just trying to merge this PR. If @kruda is not responding, can we merge this first (there are only minor issues to be fixed). And we can open some subsequent PRs to complete it. What do you think about this?

I'll remove my name from the list to avoid ambiguity. Thanks you.

**Comment by StrikerRUS** [comment]
I suppose it'll be better to continue the work (apply fixes) directly in this PR. If @kruda didn't unchecked enabled by default possibility to push into this branch by maintainers, then it'll possible.
I'd better prefer not having the repository in a undesirable state if it can be avoided (my position hasn't changed since https://github.com/microsoft/LightGBM/pull/2815#issuecomment-591030239).
https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by shiyu1994** [comment]
Oh sure. It would be the best if we can directly push to this branch. Thanks for your suggestion.

BTW, @StrikerRUS @jameslamb, I think both of you are more experienced then me in managing open-source projects. So if any of my behavior makes you confused or seems no good to the project, please feel free to correct me. Thanks!

**Comment by StrikerRUS** [comment]
@shiyu1994 Thank you for your kindness and openness! I believe everything is fine, please don't worry.

**Comment by jameslamb** [comment]
> I suppose it'll be better to continue the work (apply fixes) directly in this PR

I agree with @StrikerRUS . If someone opens a pull request into this project from their fork (even `master`), unless they've explicitly opted out, I think it's ok for maintainers to push directly to that branch to ensure the PR gets finished in an acceptable amount of time.

It's ok for you to push changes directly here @shiyu1994 . Then @StrikerRUS , I, and others can review them when you're ready.

**Comment by guolinke** [comment]
I think it is better to merge this PR into a branch of the LightGBM repo first, then we continue to develop on that branch, and merge it to the master branch when it ready.

**Comment by StrikerRUS** [comment]
@guolinke 

> I think it is better to merge this PR into a branch of the LightGBM repo first, then we continue to develop on that branch, and merge it to the master branch when it is ready.

Just created a new `mvs_dev` branch and re-targeted this PR into it.

**Comment by guolinke** [comment]
the merge is blocked by conflicts 😅 , it seems most of them are document related. @shiyu1994 @jameslamb  can you help to reslove them?

**Comment by shiyu1994** [comment]
I'm picking this up. This PR is almost ready.

**Comment by guolinke** [comment]
@shiyu1994 then we still merge it to mvs_dev branch first, then you can develop based on it?

**Comment by jameslamb** [comment]
> then we still merge it to mvs_dev branch first, then you can develop based on it?

We should definitely do that. @kruda could delete their fork or this branch at any time. Let's move merge this code to `mvs_dev` as soon as possible so that doesn't happen.

**Comment by shiyu1994** [comment]
> Let's move merge this code to mvs_dev as soon as possible

Done with that. Will work on `microsoft:mvs_dev` instead.

**Comment by StrikerRUS** [comment]
@kruda Thank you so much for your work!

**Pull request merged**
